'MSB6006: "csc.exe" exited with code 1' when analyzer ID has whitespaces

1. Create an Analyzer with following code, which just fires a diagnostic on every named type. Note that the analyzer uses an ID which has whitespace characters and a message string with multiple lines.

    ```c#
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class Analyzer2Analyzer : DiagnosticAnalyzer
    {
        public static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
            "My ID",
            @"My Title",
            @"My Message with
            multiple
            new
            lines.",
            "Category",
            DiagnosticSeverity.Error,
            true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
        }

        private void AnalyzeSymbol(SymbolAnalysisContext context)
        {
            context.ReportDiagnostic(Diagnostic.Create(Descriptor, context.Symbol.Locations[0]));
        }
    }
    ```

2. Add the analyzer nuget package reference to a C# project and build project (or invoke msbuild from command line console).
**Got:**
    ```
1>------ Rebuild All started: Project: ConsoleApplication33, Configuration: Debug Any CPU ------
1>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.CSharp.Core.targets(67,5): error MSB6006: "csc.exe" exited with code 1.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
    ```
Error list also contains the same error.
If you set the output window verbosity to detailed you can see the following:
    ```
1>  Microsoft (R) Visual C# Compiler version 1.1.0.50804
1>  Copyright (C) Microsoft Corporation. All rights reserved.
1>  
1>  Program.cs(9,11,9,18): error My ID: My Message with
1>              multiple
1>              new
1>              lines.
1>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.CSharp.Core.targets(67,5): error MSB6006: "csc.exe" exited with code 1.
1>Done executing task "Csc" -- FAILED.
1>Done building target "CoreCompile" in project "ConsoleApplication33.csproj" -- FAILED.
    ```
So, basically the C# compiler reports the diagnostic, but MSBuild parsing logic doesn't work and swallows all the compiler output.

3. Now remove the white space from analyzer ID, i.e. change it to "MyId" and regenerated the analyzer nuget package. Create a new C# project with this nuget package and build it.
**Got:**
    ```
1>------ Rebuild All started: Project: ConsoleApplication34, Configuration: Debug Any CPU ------
1>c:\users\mavasani\documents\visual studio 2015\Projects\ConsoleApplication34\ConsoleApplication34\Program.cs(9,11,9,18): error MyID: My Message with
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
    ```
MSBuild does report the diagnostic, but the message is truncated to have only the first line of the message string.


We need to fix C# and VB MSBuild tasks to parse diagnostics with:

1. ID having any characters including whitespaces, special characters, etc. A 3rd party analyzer can have any ID for their diagnostics.
2. Message with multiple lines and any characters.

--- a/src/Compilers/CSharp/Portable/Parser/CharacterInfo.cs
+++ b/src/Compilers/CSharp/Portable/Parser/CharacterInfo.cs
@@ -144,32 +144,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <param name="ch">The Unicode character.</param>
         public static bool IsIdentifierStartCharacter(char ch)
         {
-            // identifier-start-character:
-            //   letter-character
-            //   _ (the underscore character U+005F)
-
-            if (ch < 'a') // '\u0061'
-            {
-                if (ch < 'A') // '\u0041'
-                {
-                    return false;
-                }
-
-                return ch <= 'Z'  // '\u005A'
-                    || ch == '_'; // '\u005F'
-            }
-
-            if (ch <= 'z') // '\u007A'
-            {
-                return true;
-            }
-
-            if (ch <= '\u007F') // max ASCII
-            {
-                return false;
-            }
-
-            return IsLetterChar(CharUnicodeInfo.GetUnicodeCategory(ch));
+            return UnicodeCharacterUtilities.IsIdentifierStartCharacter(ch);
         }
 
         /// <summary>
@@ -178,41 +153,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <param name="ch">The Unicode character.</param>
         public static bool IsIdentifierPartCharacter(char ch)
         {
-            // identifier-part-character:
-            //   letter-character
-            //   decimal-digit-character
-            //   connecting-character
-            //   combining-character
-            //   formatting-character
-
-            if (ch < 'a') // '\u0061'
-            {
-                if (ch < 'A') // '\u0041'
-                {
-                    return ch >= '0'  // '\u0030'
-                        && ch <= '9'; // '\u0039'
-                }
-
-                return ch <= 'Z'  // '\u005A'
-                    || ch == '_'; // '\u005F'
-            }
-
-            if (ch <= 'z') // '\u007A'
-            {
-                return true;
-            }
-
-            if (ch <= '\u007F') // max ASCII
-            {
-                return false;
-            }
-
-            UnicodeCategory cat = CharUnicodeInfo.GetUnicodeCategory(ch);
-            return IsLetterChar(cat)
-                || IsDecimalDigitChar(cat)
-                || IsConnectingChar(cat)
-                || IsCombiningChar(cat)
-                || IsFormattingChar(cat);
+            return UnicodeCharacterUtilities.IsIdentifierPartCharacter(ch);
         }
 
         /// <summary>
@@ -220,26 +161,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// </summary>
         public static bool IsValidIdentifier(string name)
         {
-            if (string.IsNullOrEmpty(name))
-            {
-                return false;
-            }
-
-            if (!SyntaxFacts.IsIdentifierStartCharacter(name[0]))
-            {
-                return false;
-            }
-
-            int nameLength = name.Length;
-            for (int i = 1; i < nameLength; i++) //NB: start at 1
-            {
-                if (!SyntaxFacts.IsIdentifierPartCharacter(name[i]))
-                {
-                    return false;
-                }
-            }
-
-            return true;
+            return UnicodeCharacterUtilities.IsValidIdentifier(name);
         }
 
         /// <summary>
@@ -265,91 +187,13 @@ namespace Microsoft.CodeAnalysis.CSharp
             int nameLength = name.Length;
             for (int i = 0; i < nameLength; i++)
             {
-                if (SyntaxFacts.IsFormattingChar(name[i]))
+                if (UnicodeCharacterUtilities.IsFormattingChar(name[i]))
                 {
                     return true;
                 }
             }
 
             return false;
-        }
-
-        internal static bool IsLetterChar(UnicodeCategory cat)
-        {
-            // letter-character:
-            //   A Unicode character of classes Lu, Ll, Lt, Lm, Lo, or Nl 
-            //   A Unicode-escape-sequence representing a character of classes Lu, Ll, Lt, Lm, Lo, or Nl
-
-            switch (cat)
-            {
-                case UnicodeCategory.UppercaseLetter:
-                case UnicodeCategory.LowercaseLetter:
-                case UnicodeCategory.TitlecaseLetter:
-                case UnicodeCategory.ModifierLetter:
-                case UnicodeCategory.OtherLetter:
-                case UnicodeCategory.LetterNumber:
-                    return true;
-            }
-
-            return false;
-        }
-
-        internal static bool IsCombiningChar(UnicodeCategory cat)
-        {
-            // combining-character:
-            //   A Unicode character of classes Mn or Mc 
-            //   A Unicode-escape-sequence representing a character of classes Mn or Mc
-
-            switch (cat)
-            {
-                case UnicodeCategory.NonSpacingMark:
-                case UnicodeCategory.SpacingCombiningMark:
-                    return true;
-            }
-
-            return false;
-        }
-
-        internal static bool IsDecimalDigitChar(UnicodeCategory cat)
-        {
-            // decimal-digit-character:
-            //   A Unicode character of the class Nd 
-            //   A unicode-escape-sequence representing a character of the class Nd
-
-            return cat == UnicodeCategory.DecimalDigitNumber;
-        }
-
-        internal static bool IsConnectingChar(UnicodeCategory cat)
-        {
-            // connecting-character:  
-            //   A Unicode character of the class Pc
-            //   A unicode-escape-sequence representing a character of the class Pc
-
-            return cat == UnicodeCategory.ConnectorPunctuation;
-        }
-
-        /// <summary>
-        /// Returns true if the Unicode character is a formatting character (Unicode class Cf).
-        /// </summary>
-        /// <param name="ch">The Unicode character.</param>
-        internal static bool IsFormattingChar(char ch)
-        {
-            // There are no FormattingChars in ASCII range
-
-            return ch > 127 && IsFormattingChar(CharUnicodeInfo.GetUnicodeCategory(ch));
-        }
-
-        /// <summary>
-        /// Returns true if the Unicode character is a formatting character (Unicode class Cf).
-        /// </summary>
-        /// <param name="cat">The Unicode character.</param>
-        internal static bool IsFormattingChar(UnicodeCategory cat)
-        {
-            // formatting-character:  
-            //   A Unicode character of the class Cf
-            //   A unicode-escape-sequence representing a character of the class Cf
-
-            return cat == UnicodeCategory.Format;
         }
 
         internal static bool IsNonAsciiQuotationMark(char ch)

--- a/src/Compilers/CSharp/Portable/Parser/Lexer.cs
+++ b/src/Compilers/CSharp/Portable/Parser/Lexer.cs
@@ -1694,7 +1694,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax
                             else if (_identLen > 0 && ch > 127 && SyntaxFacts.IsIdentifierPartCharacter(ch))
                             {
                                 //// BUG 424819 : Handle identifier chars > 0xFFFF via surrogate pairs
-                                if (SyntaxFacts.IsFormattingChar(ch))
+                                if (UnicodeCharacterUtilities.IsFormattingChar(ch))
                                 {
                                     if (isEscaped)
                                     {
@@ -1982,7 +1982,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax
                             else if (_identLen > 0 && consumedChar > 127 && SyntaxFacts.IsIdentifierPartCharacter(consumedChar))
                             {
                                 //// BUG 424819 : Handle identifier chars > 0xFFFF via surrogate pairs
-                                if (SyntaxFacts.IsFormattingChar(consumedChar))
+                                if (UnicodeCharacterUtilities.IsFormattingChar(consumedChar))
                                 {
                                     continue; // Ignore formatting characters
                                 }

--- a/src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs
@@ -984,6 +984,21 @@ public class B
             }
         }
 
+        [Fact, WorkItem(4376, "https://github.com/dotnet/roslyn/issues/4376")]
+        public void TestReportingDiagnosticWithInvalidId()
+        {
+            string source = @"";
+            var analyzers = new DiagnosticAnalyzer[] { new AnalyzerWithInvalidDiagnosticId() };
+            string message = new ArgumentException(string.Format(CodeAnalysisResources.InvalidDiagnosticIdReported, AnalyzerWithInvalidDiagnosticId.Descriptor.Id), "diagnostic").Message;
+
+            CreateCompilationWithMscorlib45(source)
+                .VerifyDiagnostics()
+                .VerifyAnalyzerDiagnostics(analyzers, null, null, logAnalyzerExceptionAsDiagnostics: true,
+                     expected: Diagnostic("AD0001")
+                     .WithArguments("Microsoft.CodeAnalysis.CommonDiagnosticAnalyzers+AnalyzerWithInvalidDiagnosticId", "System.ArgumentException", message)
+                     .WithLocation(1, 1));
+        }
+
         [Fact, WorkItem(1473, "https://github.com/dotnet/roslyn/issues/1473")]
         public void TestReportingNotConfigurableDiagnostic()
         {

--- a/src/Compilers/Core/Portable/CodeAnalysis.csproj
+++ b/src/Compilers/Core/Portable/CodeAnalysis.csproj
@@ -51,6 +51,7 @@
     <Compile Include="AssemblyUtilities.cs" />
     <Compile Include="Binding\AbstractLookupSymbolsInfo.cs" />
     <Compile Include="CaseInsensitiveComparison.cs" />
+    <Compile Include="UnicodeCharacterUtilities.cs" />
     <Compile Include="CodeAnalysisResources.Designer.cs">
       <AutoGen>True</AutoGen>
       <DesignTime>True</DesignTime>

--- a/src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs
+++ b/src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs
@@ -577,6 +577,15 @@ namespace Microsoft.CodeAnalysis {
         }
         
         /// <summary>
+        ///   Looks up a localized string similar to Reported diagnostic has an ID &apos;{0}&apos;, which is not a valid identifier..
+        /// </summary>
+        internal static string InvalidDiagnosticIdReported {
+            get {
+                return ResourceManager.GetString("InvalidDiagnosticIdReported", resourceCulture);
+            }
+        }
+        
+        /// <summary>
         ///   Looks up a localized string similar to Invalid hash..
         /// </summary>
         internal static string InvalidHash {

--- a/src/Compilers/Core/Portable/CodeAnalysisResources.resx
+++ b/src/Compilers/Core/Portable/CodeAnalysisResources.resx
@@ -401,6 +401,9 @@
   <data name="UnsupportedDiagnosticReported" xml:space="preserve">
     <value>Reported diagnostic with ID '{0}' is not supported by the analyzer.</value>
   </data>
+  <data name="InvalidDiagnosticIdReported" xml:space="preserve">
+    <value>Reported diagnostic has an ID '{0}', which is not a valid identifier.</value>
+  </data>
   <data name="NoBinderException" xml:space="preserve">
     <value>Cannot deserialize type '{0}', no binder supplied.</value>
   </data>

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs
@@ -860,6 +860,12 @@ namespace Microsoft.CodeAnalysis.Diagnostics
         {
             Debug.Assert(_isCompilerAnalyzer != null);
 
+            if (diagnostic is DiagnosticWithInfo)
+            {
+                // Compiler diagnostic
+                return true;
+            }
+
             return _analyzerManager.IsSupportedDiagnostic(analyzer, diagnostic, _isCompilerAnalyzer, this);
         }
 

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContextHelpers.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContextHelpers.cs
@@ -27,6 +27,12 @@ namespace Microsoft.CodeAnalysis.Diagnostics
 
         internal static void VerifyArguments(Diagnostic diagnostic, Func<Diagnostic, bool> isSupportedDiagnostic)
         {
+            if (diagnostic is DiagnosticWithInfo)
+            {
+                // Compiler diagnostic, skip validations.
+                return;
+            }
+
             if (diagnostic == null)
             {
                 throw new ArgumentNullException(nameof(diagnostic));
@@ -36,6 +42,14 @@ namespace Microsoft.CodeAnalysis.Diagnostics
             {
                 throw new ArgumentException(string.Format(CodeAnalysisResources.UnsupportedDiagnosticReported, diagnostic.Id), nameof(diagnostic));
             }
+
+            if (!UnicodeCharacterUtilities.IsValidIdentifier(diagnostic.Id))
+            {
+                // Disallow invalid diagnostic IDs.
+                // Note that the parsing logic in Csc/Vbc MSBuild tasks to decode command line compiler output relies on diagnostics having a valid ID.
+                // See https://github.com/dotnet/roslyn/issues/4376 for details.
+                throw new ArgumentException(string.Format(CodeAnalysisResources.InvalidDiagnosticIdReported, diagnostic.Id), nameof(diagnostic));
+            } 
         }
 
         private static void VerifyAction<TContext>(Action<TContext> action)

--- a/src/Compilers/Core/Portable/UnicodeCharacterUtilities.cs
+++ b/src/Compilers/Core/Portable/UnicodeCharacterUtilities.cs
@@ -1,0 +1,191 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.Globalization;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Defines a set of helper methods to classify Unicode characters.
+    /// </summary>
+    internal static partial class UnicodeCharacterUtilities
+    {
+        public static bool IsIdentifierStartCharacter(char ch)
+        {
+            // identifier-start-character:
+            //   letter-character
+            //   _ (the underscore character U+005F)
+
+            if (ch < 'a') // '\u0061'
+            {
+                if (ch < 'A') // '\u0041'
+                {
+                    return false;
+                }
+
+                return ch <= 'Z'  // '\u005A'
+                    || ch == '_'; // '\u005F'
+            }
+
+            if (ch <= 'z') // '\u007A'
+            {
+                return true;
+            }
+
+            if (ch <= '\u007F') // max ASCII
+            {
+                return false;
+            }
+
+            return IsLetterChar(CharUnicodeInfo.GetUnicodeCategory(ch));
+        }
+        
+        /// <summary>
+         /// Returns true if the Unicode character can be a part of an identifier.
+         /// </summary>
+         /// <param name="ch">The Unicode character.</param>
+        public static bool IsIdentifierPartCharacter(char ch)
+        {
+            // identifier-part-character:
+            //   letter-character
+            //   decimal-digit-character
+            //   connecting-character
+            //   combining-character
+            //   formatting-character
+
+            if (ch < 'a') // '\u0061'
+            {
+                if (ch < 'A') // '\u0041'
+                {
+                    return ch >= '0'  // '\u0030'
+                        && ch <= '9'; // '\u0039'
+                }
+
+                return ch <= 'Z'  // '\u005A'
+                    || ch == '_'; // '\u005F'
+            }
+
+            if (ch <= 'z') // '\u007A'
+            {
+                return true;
+            }
+
+            if (ch <= '\u007F') // max ASCII
+            {
+                return false;
+            }
+
+            UnicodeCategory cat = CharUnicodeInfo.GetUnicodeCategory(ch);
+            return IsLetterChar(cat)
+                || IsDecimalDigitChar(cat)
+                || IsConnectingChar(cat)
+                || IsCombiningChar(cat)
+                || IsFormattingChar(cat);
+        }
+
+        /// <summary>
+        /// Check that the name is a valid Unicode identifier.
+        /// </summary>
+        public static bool IsValidIdentifier(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                return false;
+            }
+
+            if (!IsIdentifierStartCharacter(name[0]))
+            {
+                return false;
+            }
+
+            int nameLength = name.Length;
+            for (int i = 1; i < nameLength; i++) //NB: start at 1
+            {
+                if (!IsIdentifierPartCharacter(name[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsLetterChar(UnicodeCategory cat)
+        {
+            // letter-character:
+            //   A Unicode character of classes Lu, Ll, Lt, Lm, Lo, or Nl 
+            //   A Unicode-escape-sequence representing a character of classes Lu, Ll, Lt, Lm, Lo, or Nl
+
+            switch (cat)
+            {
+                case UnicodeCategory.UppercaseLetter:
+                case UnicodeCategory.LowercaseLetter:
+                case UnicodeCategory.TitlecaseLetter:
+                case UnicodeCategory.ModifierLetter:
+                case UnicodeCategory.OtherLetter:
+                case UnicodeCategory.LetterNumber:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsCombiningChar(UnicodeCategory cat)
+        {
+            // combining-character:
+            //   A Unicode character of classes Mn or Mc 
+            //   A Unicode-escape-sequence representing a character of classes Mn or Mc
+
+            switch (cat)
+            {
+                case UnicodeCategory.NonSpacingMark:
+                case UnicodeCategory.SpacingCombiningMark:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool IsDecimalDigitChar(UnicodeCategory cat)
+        {
+            // decimal-digit-character:
+            //   A Unicode character of the class Nd 
+            //   A unicode-escape-sequence representing a character of the class Nd
+
+            return cat == UnicodeCategory.DecimalDigitNumber;
+        }
+
+        private static bool IsConnectingChar(UnicodeCategory cat)
+        {
+            // connecting-character:  
+            //   A Unicode character of the class Pc
+            //   A unicode-escape-sequence representing a character of the class Pc
+
+            return cat == UnicodeCategory.ConnectorPunctuation;
+        }
+
+        /// <summary>
+        /// Returns true if the Unicode character is a formatting character (Unicode class Cf).
+        /// </summary>
+        /// <param name="ch">The Unicode character.</param>
+        internal static bool IsFormattingChar(char ch)
+        {
+            // There are no FormattingChars in ASCII range
+
+            return ch > 127 && IsFormattingChar(CharUnicodeInfo.GetUnicodeCategory(ch));
+        }
+
+        /// <summary>
+        /// Returns true if the Unicode character is a formatting character (Unicode class Cf).
+        /// </summary>
+        /// <param name="cat">The Unicode character.</param>
+        private static bool IsFormattingChar(UnicodeCategory cat)
+        {
+            // formatting-character:  
+            //   A Unicode character of the class Cf
+            //   A unicode-escape-sequence representing a character of the class Cf
+
+            return cat == UnicodeCategory.Format;
+        }
+    }
+}

--- a/src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs
+++ b/src/EditorFeatures/CSharpTest/Diagnostics/Suppression/SuppressionTests.cs
@@ -513,52 +513,12 @@ using System;
                 [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSuppression)]
                 public void TestDiagnosticWithBadIdSuppressed()
                 {
-                    Test(
+                    // Diagnostics with bad/invalid ID are not reported.
+                    TestMissing(
             @"
 using System;
 
 [|class Class|]
-{
-    int Method()
-    {
-        int x = 0;
-    }
-}",
-            @"
-using System;
-
-#pragma warning disable @~DiagnosticWithBadId // DiagnosticWithBadId
-class Class
-#pragma warning restore @~DiagnosticWithBadId // DiagnosticWithBadId
-{
-    int Method()
-    {
-        int x = 0;
-    }
-}");
-
-                    // Verify that the original suppression doesn't really work and that the diagnostic can be suppressed again.
-                    Test(
-            @"
-using System;
-
-#pragma warning disable @~DiagnosticWithBadId // DiagnosticWithBadId
-[|class Class|]
-#pragma warning restore @~DiagnosticWithBadId // DiagnosticWithBadId
-{
-    int Method()
-    {
-        int x = 0;
-    }
-}",
-            @"
-using System;
-
-#pragma warning disable @~DiagnosticWithBadId // DiagnosticWithBadId
-#pragma warning disable @~DiagnosticWithBadId // DiagnosticWithBadId
-class Class
-#pragma warning restore @~DiagnosticWithBadId // DiagnosticWithBadId
-#pragma warning restore @~DiagnosticWithBadId // DiagnosticWithBadId
 {
     int Method()
     {

--- a/src/EditorFeatures/VisualBasicTest/Diagnostics/Suppression/SuppressionTests.vb
+++ b/src/EditorFeatures/VisualBasicTest/Diagnostics/Suppression/SuppressionTests.vb
@@ -715,6 +715,7 @@ End Class]]>
                 <WorkItem(730770)>
                 <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSuppression)>
                 Public Sub TestDiagnosticWithBadIdSuppressed()
+                    ' Diagnostics with bad/invalid ID are not reported.
                     Dim source = <![CDATA[
 Imports System
 
@@ -722,39 +723,8 @@ Imports System
     Sub Method()
     End Sub
 End Class]]>
-                    Dim expected = <![CDATA[
-Imports System
 
-#Disable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-Class C
-#Enable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-    Sub Method()
-    End Sub
-End Class]]>
-                    Test(source.Value, expected.Value)
-
-                    Dim source2 = <![CDATA[
-Imports System
-
-#Disable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-[|Class C|]
-#Enable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-    Sub Method()
-    End Sub
-End Class]]>
-                    Dim expected2 = <![CDATA[
-Imports System
-
-#Disable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-#Disable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-Class C
-#Enable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-#Enable Warning #$DiagnosticWithBadId ' DiagnosticWithBadId
-    Sub Method()
-    End Sub
-End Class]]>
-                    ' Verify that the original suppression doesn't really work and that the diagnostic can be suppressed again.
-                    Test(source2.Value, expected2.Value)
+                    TestMissing(source.Value)
                 End Sub
             End Class
 

--- a/src/ExpressionEvaluator/CSharp/Source/ResultProvider/NetFX20/CSharpResultProvider.NetFX20.csproj
+++ b/src/ExpressionEvaluator/CSharp/Source/ResultProvider/NetFX20/CSharpResultProvider.NetFX20.csproj
@@ -19,6 +19,9 @@
     <RestorePackages>true</RestorePackages>
   </PropertyGroup>
   <ItemGroup Label="Linked Files">
+    <Compile Include="..\..\..\..\..\Compilers\Core\Portable\UnicodeCharacterUtilities.cs">
+      <Link>Compiler\Parser\UnicodeCharacterUtilities.cs</Link>
+    </Compile>
     <Compile Include="..\..\..\..\..\Compilers\CSharp\Portable\Parser\CharacterInfo.cs">
       <Link>Compiler\Parser\CharacterInfo.cs</Link>
     </Compile>

--- a/src/ExpressionEvaluator/CSharp/Source/ResultProvider/Portable/CSharpResultProvider.Portable.csproj
+++ b/src/ExpressionEvaluator/CSharp/Source/ResultProvider/Portable/CSharpResultProvider.Portable.csproj
@@ -18,6 +18,9 @@
     <ProjectTypeGuids>{786C830F-07A1-408B-BD7F-6EE04809D6DB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
   </PropertyGroup>
   <ItemGroup Label="Linked Files">
+    <Compile Include="..\..\..\..\..\Compilers\Core\Portable\UnicodeCharacterUtilities.cs">
+      <Link>Compiler\Parser\UnicodeCharacterUtilities.cs</Link>
+    </Compile>
     <Compile Include="..\..\..\..\..\Compilers\CSharp\Portable\Parser\CharacterInfo.cs">
       <Link>Compiler\Parser\CharacterInfo.cs</Link>
     </Compile>

--- a/src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs
+++ b/src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs
@@ -299,6 +299,25 @@ namespace Microsoft.CodeAnalysis
         }
 
         [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+        public sealed class AnalyzerWithInvalidDiagnosticId : DiagnosticAnalyzer
+        {
+            public static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
+                "Invalid ID",
+                "Title1",
+                "Message1",
+                "Category1",
+                defaultSeverity: DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+            public override void Initialize(AnalysisContext context)
+            {
+                context.RegisterCompilationAction(compilationContext =>
+                    compilationContext.ReportDiagnostic(Diagnostic.Create(Descriptor, Location.None)));
+            }
+        }
+
+        [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
         public sealed class AnalyzerThatThrowsInGetMessage : DiagnosticAnalyzer
         {
             public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
@@ -341,7 +360,7 @@ namespace Microsoft.CodeAnalysis
         public class HiddenDiagnosticsCompilationAnalyzer : DiagnosticAnalyzer
         {
             public static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
-                "ID 1000",
+                "ID1000",
                 "Description1",
                 string.Empty,
                 "Analysis",

